/*
 * Copyright 2024 The PhoenixOS Authors. All rights reserved.
 * 
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

syntax = "proto3";

package pos_protobuf;


message Bin_POSHandleView {
    uint32 resource_type_id = 1;
    uint64 id = 2;
    uint64 param_index = 3;
    uint64 offset = 4;
}


/*!
 *  \brief  binary format of POSAPIParam
 */
message Bin_POSAPIParam {
    uint64 size = 1;
    bytes state = 2;
}


/*!
 *  \brief  binary format of POSAPIContext
 */
message Bin_POSAPIContext {
    uint64 id = 1;
    uint64 api_id = 2;
    bool has_return = 3;
    repeated Bin_POSHandleView input_handle_views = 4;
    repeated Bin_POSHandleView output_handle_views = 5;
    repeated Bin_POSHandleView create_handle_views = 6;
    repeated Bin_POSHandleView delete_handle_views = 7;
    repeated Bin_POSHandleView inout_handle_views = 8;
    uint64 create_tick = 9;
    uint64 return_tick = 10;
    uint64 parser_s_tick = 11;
    uint64 parser_e_tick = 12;
    uint64 worker_s_tick = 13;
    uint64 worker_e_tick = 14;
    repeated Bin_POSAPIParam params = 15;
    uint64 retval_size = 16;
}
